consensus/polybft/consensus_runtime.go

```diff
@@ -417,81 +440,68 @@ func (c *consensusRuntime) FSM() error {
 }
 
 // restartEpoch resets the previously run epoch and moves to the next one
-func (c *consensusRuntime) restartEpoch(header *types.Header) error {
+// returns *epochMetadata different from nil if the lastEpoch is not the current one and everything was successful
+func (c *consensusRuntime) restartEpoch(header *types.Header, lastEpoch *epochMetadata) (*epochMetadata, error) {
 	systemState, err := c.getSystemState(header)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	epochNumber, err := systemState.GetEpoch()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	lastEpoch := c.getEpoch()
 	if lastEpoch != nil {
-		// Epoch might be already in memory, if its the same number do nothing.
+		// Epoch might be already in memory, if its the same number do nothing -> just return provided last one
 		// Otherwise, reset the epoch metadata and restart the async services
 		if lastEpoch.Number == epochNumber {
-			return nil
+			return lastEpoch, nil
 		}
 	}
 
 	validatorSet, err := c.config.polybftBackend.GetValidators(header.Number, nil)
 	if err != nil {
-		return err
-	}
-
-	epoch := &epochMetadata{
-		Number:         epochNumber,
-		LastCheckpoint: 0,
-		Blocks:         []*types.Block{},
-		Validators:     validatorSet,
+		return nil, err
 	}
 
 	if err := c.state.cleanEpochsFromDB(); err != nil {
 		c.logger.Error("Could not clean previous epochs from db.", "error", err)
 	}
 
-	if err := c.state.insertEpoch(epoch.Number); err != nil {
-		return fmt.Errorf("an error occurred while inserting new epoch in db. Reason: %w", err)
+	if err := c.state.insertEpoch(epochNumber); err != nil {
+		return nil, fmt.Errorf("an error occurred while inserting new epoch in db. Reason: %w", err)
 	}
 
+	var commitment *Commitment = nil
+
 	// create commitment for state sync events
 	if c.IsBridgeEnabled() {
 		nextCommittedIndex, err := systemState.GetNextCommittedIndex()
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		commitment, err := c.buildCommitment(epochNumber, nextCommittedIndex)
+		commitment, err = c.buildCommitment(epochNumber, nextCommittedIndex)
 		if err != nil {
-			return err
+			return nil, err
 		}
-
-		epoch.Commitment = commitment
 	}
 
-	c.lock.Lock()
-	c.epoch = epoch
-	c.lastBuiltBlock = header
-
-	proposerSnapshot := c.proposerSnapshot.Copy()
-	if err := c.updateProposerSnapshot(proposerSnapshot, header.Number); err != nil {
-		c.logger.Warn("Could not update proposer calculator", "err", err)
-	}
-
-	c.proposerSnapshot = proposerSnapshot
-	c.lock.Unlock()
-
 	c.logger.Info(
 		"restartEpoch",
 		"block number", header.Number,
 		"epoch", epochNumber,
 		"validators", validatorSet.Len(),
 	)
 
-	return nil
+	return &epochMetadata{
+		Number:         epochNumber,
+		LastCheckpoint: 0,
```

Comment:
You don't need to set this `LastCheckpoint` and `Blocks` fields, they are removed throuygh another PR, plus, they are not used anywhere.

Reply:
fixed https://github.com/0xPolygon/polygon-edge/pull/1047/commits/9ad952c1a99a2bacf73e1e6a2146d02f22021ec2